encoding.bs

```diff
@@ -1018,61 +1018,56 @@ run these steps:
  <li><p>If <var>buffer</var> does not match 0xEF 0xBB 0xBF, <a>prepend</a> <var>buffer</var> to
  <var>ioQueue</var>.
 
+ <li><p><a>Run</a> <a>UTF-8</a>'s <a for=/>decoder</a> with <var>ioQueue</var> and
+ <var>output</var>.
 
- <li><p>Let <var>output</var> be the <a for=/>I/O queue</a> of scalar values
- « <a>end-of-queue</a> ».
-
- <li><p><a>Run</a> <a>UTF-8</a>'s
- <a for=/>decoder</a> with <var>ioQueue</var> and <var>output</var>.
+ <li><p><a for="I/O queue">Push</a> <a>end-of-queue</a> into <var>output</var>.
 
  <li><p>Return <var>output</var>.
 </ol>
 
-<p>To <dfn export>UTF-8 decode without BOM</dfn> an I/O queue of bytes
-<var>ioQueue</var>, run these steps:
+<p>To <dfn export>UTF-8 decode without BOM</dfn> an I/O queue of bytes <var>ioQueue</var> given an
+optional I/O queue of scalar values <var>output</var> (default « »), run these steps:
 
 <ol>
- <li><p>Let <var>output</var> be the <a for=/>I/O queue</a> of scalar values
- « <a>end-of-queue</a> ».
+ <li><p><a>Run</a> <a>UTF-8</a>'s <a for=/>decoder</a> with <var>ioQueue</var> and
+ <var>output</var>.
 
- <li><p><a>Run</a> <a>UTF-8</a>'s
- <a for=/>decoder</a> with <var>ioQueue</var> and <var>output</var>.
+ <li><p><a for="I/O queue">Push</a> <a>end-of-queue</a> into <var>output</var>.
 
  <li><p>Return <var>output</var>.
 </ol>
 
-<p>To <dfn export>UTF-8 decode without BOM or fail</dfn> an I/O queue of bytes
-<var>ioQueue</var>, run these steps:
+<p>To <dfn export>UTF-8 decode without BOM or fail</dfn> an I/O queue of bytes <var>ioQueue</var>
+given an optional I/O queue of scalar values <var>output</var> (default « »), run these steps:
 <!-- Needed by https://tools.ietf.org/html/rfc6455#section-8.1 and
      https://webassembly.github.io/spec/js-api/#dom-module-customsections-moduleobject-sectionname
      -->
 
 <ol>
- <li><p>Let <var>output</var> be the <a for=/>I/O queue</a> of scalar values
- « <a>end-of-queue</a> ».
-
- <li><p>Let <var>potentialError</var> be the result of <a>running</a>
- <a>UTF-8</a>'s <a for=/>decoder</a> with <var>ioQueue</var>, <var>output</var>,
- and "<code>fatal</code>".
+ <li><p>Let <var>potentialError</var> be the result of <a>running</a> <a>UTF-8</a>'s
+ <a for=/>decoder</a> with <var>ioQueue</var>, <var>output</var>, and "<code>fatal</code>".
 
  <li><p>If <var>potentialError</var> is <a>error</a>, return failure.
 
+ <li><p><a for="I/O queue">Push</a> <a>end-of-queue</a> into <var>output</var>.
```

Comment:
Do we push in the failure case too? For the case of an XML parser with the tokenizer running in parallel from the decoding, we wouldn't want the tokenizer to process the end-of-queue token before the document is blown away, even in cases where the tokenizer wouldn't error (say, inside a comment after the root element's end tag).

Reply:
So currently "Hooks for standards" doesn't really cover the XML case. That needs a variant of "decode" that captures the output of "run" and also passes fatal as error mode in. I can file a follow-up issue for XML if you agree.

Reply:
Making a hook for XML, that'd be fine by me, but I don't think it would be needed until and unless there's an algorithmic XML parser defined somewhere, which seems unlikely to happen anytime soon.